tap_hubspot/__init__.py

```diff
@@ -1131,6 +1220,67 @@ class Stream:
     Stream('engagements', sync_engagements, ["engagement_id"], 'lastUpdated', 'FULL_TABLE')
 ]
 
+# pylint: disable=inconsistent-return-statements
+def generate_custom_streams(mode, catalog=None):
+    """
+    - In DISCOVER mode, fetch the custom schema from the API endpoint and set the schema for the custom objects.
+    - In SYNC mode, extend STREAMS for the custom objects.
+
+    Args:
+        mode (str): The mode indicating whether to DISCOVER or SYNC custom streams.
+        catalog (dict): The catalog containing stream information.
+
+    Returns:
+        List[dict] or List[str]: Returns list of custom streams (contains dictionary) in DISCOVER mode and a list of custom object names in SYNC mode.
+    """
+    custom_objects_schema_url = get_url("custom_objects_schema")
+    if mode == "DISCOVER":
+        custom_streams = []
+        # Load Hubspot's shared schemas
+        refs = load_shared_schema_refs()
+        try:
+            for custom_object in gen_request_custom_objects("custom_objects_schema", custom_objects_schema_url, {}, 'results', "paging"):
+                stream_id = custom_object["name"]
+                schema = utils.load_json(get_abs_path('schemas/shared/custom_objects.json'))
+                custom_schema = parse_custom_schema(stream_id, custom_object["properties"], is_custom_object=True)
+                schema["properties"]["properties"] = {
+                    "type": "object",
+                    "properties": custom_schema,
+                }
+
+                # Move properties to top level
+                custom_schema_top_level = {'property_{}'.format(k): v for k, v in custom_schema.items()}
+                schema['properties'].update(custom_schema_top_level)
+
+                final_schema = singer.resolve_schema_references(schema, refs)
+                custom_streams.append({"stream": Stream(stream_id, sync_custom_object_records, ['id'], 'updatedAt', 'INCREMENTAL'),
+                                       "schema": final_schema})
+
+        except SourceUnavailableException as ex:
+            warning_message = str(ex).replace(CONFIG['access_token'], 10 * '*')
+            LOGGER.warning(warning_message)
+        return custom_streams
+
+    elif mode == "SYNC":
+        custom_objects = [custom_object["name"] for custom_object in gen_request_custom_objects("custom_objects_schema", custom_objects_schema_url, {}, 'results', "paging")]
+        for stream in catalog["streams"]:
+            if stream["tap_stream_id"] in custom_objects:
+                STREAMS.append(Stream(stream["tap_stream_id"], sync_custom_object_records, ['id'], 'updatedAt', 'INCREMENTAL'))
+        return custom_objects
```

Comment:
I don't have a great idea of what it would look like, but I feel like this could be made into two separate functions to keep discovery functionality inside the discover_schemas() function?

I suppose that might not be worth the effort given that this tap has everything inside init instead of separate sync / discover files, but I felt like it was worth thinking about at least

Reply:
Yeah, I thought of extending the existing function - discover_schemas() but it was looking complex, therefore, I created a new function. Maybe I can put it in the backlog and take care while refactoring the code (main motive is to remove the redundant code.)

Reply:
👍